src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using Bicep.Core.Diagnostics;
+using Bicep.Core.SemanticModel;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.TypeSystem
+{
+    /// <summary>
+    /// Visitor used to collect errors caused by property assignments to run-time variables,
+    /// currently, we only run this on the "name" property of resources and modules
+    /// </summary>
+    public sealed class DeployTimeConstantVisitor : SyntaxVisitor
+    {
+
+        private readonly SemanticModel.SemanticModel model;
+        private readonly IDiagnosticWriter diagnosticWriter;
+
+        // used for logging the property when we detect a run time property
+        private string currentSymbol;
+        // whether a runtimeValue is allowed for the currentSymbol
+        private bool runtimeValueAllowed;
+
+        private DeployTimeConstantVisitor(SemanticModel.SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            this.model = model;
+            this.diagnosticWriter = diagnosticWriter;
+            this.runtimeValueAllowed = true;
+            this.currentSymbol = "";
+        }
+
+        // entry point for this visitor
+        public static void ValidateDeployTimeConstants(SemanticModel.SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            new DeployTimeConstantVisitor(model, diagnosticWriter).Visit(model.Root.Syntax);
+        }
+
+        public override void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
+        {
+            this.runtimeValueAllowed = false;
+            base.VisitResourceDeclarationSyntax(syntax);
+            this.runtimeValueAllowed = true;
+        }
+
+        public override void VisitModuleDeclarationSyntax(ModuleDeclarationSyntax syntax)
+        {
+            this.runtimeValueAllowed = false;
+            base.VisitModuleDeclarationSyntax(syntax);
+            this.runtimeValueAllowed = true;
+        }
+
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+        {
+            bool priorRuntimeValueAllowed = this.runtimeValueAllowed;
+            // runtimeValueAllowed should only be false if we are going through a module or resource's objectPropertySyntax
+            if (!this.runtimeValueAllowed && syntax.Key is IdentifierSyntax keyIdentifier) 
+            {
+                this.currentSymbol = keyIdentifier.IdentifierName;
+                // right now we only check for the name property (using resourceNamePropertyName but the name is same for modules)
+                if (!keyIdentifier.IdentifierName.Equals(LanguageConstants.ResourceNamePropertyName, System.StringComparison.Ordinal))
```

Comment:
>System.StringComparison.Ordinal [](start = 101, length = 31)

If you're comparing strings that are Bicep identifiers, please use the `IdentifierComparer` from `LanguageConstants`. Functionally, it's identical, but serves as an easy way to identify all the places where identifiers are compared by doing "find refs" on it.